packages/kad-dht/src/reprovider.ts

```diff
@@ -0,0 +1,226 @@
+   * them if the provider is us and the expiry is within the reprovide window.
+   */
+  private async cleanUp (): Promise<void> {
+    const release = await this.lock.writeLock()
+
+    try {
+      this.safeDispatchEvent('reprovide:start')
+
+      // Get all provider entries from the datastore
+      for await (const entry of this.datastore.query({
+        prefix: this.datastorePrefix
+      })) {
+        try {
+          // Add a delete to the batch for each expired entry
+          const { cid, peerId } = parseProviderKey(entry.key)
+          const created = readProviderTime(entry.value).getTime()
+          const expires = created + this.validity
+          const now = Date.now()
+          const expired = now > expires
+
+          this.log.trace('comparing: %d < %d = %s %s', created, now - this.validity, expired, expired ? '(expired)' : '')
+
+          // delete the record if it has expired
+          if (expired) {
+            await this.datastore.delete(entry.key)
+          }
+
+          // if the provider is us and we are within the reprovide threshold,
+          // reprovide the record
+          if (this.peerId.equals(peerId) && (now - expires) < this.reprovideThreshold) {
+            this.queueReprovide(cid)
+              .catch(err => {
+                this.log.error('could not reprovide %c - %e', cid, err)
+              })
+          }
+        } catch (err: any) {
+          this.log.error('error processing datastore key %s - %e', entry.key, err.message)
+        }
+      }
+
+      this.log('reprovide/cleanup successful')
+    } finally {
+      release()
+      this.safeDispatchEvent('reprovide:end')
+
+      if (this.running) {
+        this.timeout = setTimeout(() => {
+          this.cleanUp().catch(err => {
+            this.log.error('error running re-provide - %e', err)
+          })
+        }, this.interval)
+      }
+    }
+  }
+
+  private async queueReprovide (cid: CID): Promise<void> {
+    if (!this.running) {
+      return
+    }
+
+    this.log.trace('waiting for queue capacity before adding %c to re-provide queue', cid)
+    await this.reprovideQueue.onSizeLessThan(this.maxQueueSize)
+
+    const existingJob = this.reprovideQueue.queue.find(job => job.options.cid.equals(cid))
+
+    if (existingJob != null) {
+      this.log.trace('not adding %c to re-provide queue - already in queue', cid)
+      return existingJob.join()
+    }
+
+    this.log.trace('adding %c to re-provide queue', cid)
+
+    this.reprovideQueue.add(async (options): Promise<void> => {
+      options.signal?.throwIfAborted()
+
+      if (!this.running) {
+        return
+      }
+
+      this.log.trace('re-providing %c', cid)
+
+      // use adaptive timeout
+      const signal = this.reprovideTimeout.getTimeoutSignal(options)
+
+      try {
+        // reprovide
+        await this.reprovide(options.cid, options)
+      } finally {
+        this.reprovideTimeout.cleanUp(signal)
```

Comment:
This runs irrespective of whether the reprovide for the CID succeeds. Is that a problem?